### PR TITLE
Add command for reading a json-based DB description

### PR DESCRIPTION
**Pull request by scheibelp**
Adds `spack external read-cray-manifest`, which reads a json file that describes a set of package DAGs. The parsed results are stored directly in the database. A user can see these installed specs with `spack find` (like any installed spec). The easiest way to use them right now as dependencies is to run `spack spec ... ^/hash-of-external-package`.

Changes include:

* `spack external read-cray-manifest --file <path/to/file>` will add all specs described in the file to Spack's installation DB and will also install described compilers to the compilers configuration (the expected format of the file is described in this PR as well including examples of the file)
* Database records now may include an "origin" (the command added in this PR registers the origin as "external-db"). In the future, it is assumed users may want to be able to treat installs registered with this command differently (e.g. they may want to uninstall all specs added with this command)
* Hash properties are now always preserved when copying specs if the source spec is concrete
  * I don't think the hashes of installed-and-concrete specs should change and this was the easiest way to handle that
  * also specs that are concrete preserve their `.normal` property when copied (external specs may mention compilers that are not registered, and without this change they would fail in `normalize` when calling `validate_or_raise`)
  * it might be this should only be the case if the spec was installed

TODOs

- [x] Improve testing
- [x] Specifically mark DB records added with this command (so that users can do something like "uninstall all packages added with `spack read-external-db`)
  * This is now possible with `spack uninstall --all --origin=external-db` (this will remove all specs added from manifest files)
- [x] Strip variants that are listed in json entries but don't actually exist for the package

### CODE REVIEW

**Comment by haampie** [comment]
What's the status on this @scheibelp? Would be nice to try it

File: lib/spack/spack/cray_manifest.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        except Exception:
+            tty.warn("Could not parse entry: " + str(entry))
+
+    for entry in filter(lambda x: 'dependencies' in x, entries):
+        dependencies = entry['dependencies']
+        for name, properties in dependencies.items():
+            dep_hash = properties['hash']
+            deptypes = properties['type']
+            if dep_hash in spec_dict:
+                parent_spec = spec_dict[entry['hash']]
```
Comment (haampie) [inline lib/spack/spack/cray_manifest.py:136]: this doesn't work when `entry` is skipped above because of UnknownPackageError

**Comment by haampie** [comment]
This generally works (rebased it here https://github.com/haampie/spack/tree/features/read-db-description), but a couple issues:

I'm running Spack with `unset MODULEPATH` / a non-login shell, so that it reports `platform=linux` instead of `platform=cray`, but the imported database is all `platform=cray`, so --reuse does not work. Also the OS in the db is `sles15sp1` but spack thinks the host OS is `sles15`. Maybe we should use the host platform and distro when importing?

**Comment by haampie** [comment]
Hm, I doesn't seem like --reuse can be made to work, I've tried adding the missing version to cray-mpich, made sure the root spec has the exact same arch triplet, but no success, it concretizes to its own cray-mpich.

Seems like it only works with ^/<hash>, which is not great user experience

**Comment by haampie** [comment]
If I put it in an environment I get `==> Error: Cannot depend on 'libfabric@1.11.0.4.83%gcc@7.5.0 fabrics=verbs arch=linux-sles15-x86_64 ^libnl@3.3.0%gcc@7.5.0 arch=linux-sles15-x86_64 ^rdma-core@51mlnx1%gcc@7.5.0 arch=linux-sles15-x86_64' twice`

**Comment by haampie** [comment]
Rebased this on develop, original PR is here https://github.com/haampie/spack/tree/features/read-db-description-pre-rebase in case I screwed up something.

**Comment by haampie** [comment]
@scheibelp can you reproduce the os name issue I had? I think it's an issue to be fixed in spack platform detection, since sles15sp3 will ship a newer glibc, so the sp3 bit should go into the target string.

**Comment by scheibelp** [comment]
> I'm running Spack with unset MODULEPATH / a non-login shell, so that it reports platform=linux instead of platform=cray, but the imported database is all platform=cray, so --reuse does not work. Also the OS in the db is sles15sp1 but spack thinks the host OS is sles15. Maybe we should use the host platform and distro when importing?

> Seems like it only works with ^/, which is not great user experience

I think the primary goal for this PR is to have this work with `^/`. One issue with overriding the OS/platform is when dealing with cross-compiled systems: this shouldn't be a problem when overriding the platform but could be an issue when overriding the OS. For the OS, I can try to coordinate with locking that down to a fixed set of values (or possibly just designating between frontend and backend).

File: lib/spack/spack/spec.py [hunk 2]
```diff
@@ -2979,8 +2979,8 @@ def normalize(self, force=False, tests=False, user_spec_deps=None):
                 "Attempting to normalize anonymous spec")
 
         # Set _normal and _concrete to False when forced
-        if force:
-            self._mark_concrete(False)
+        if force and not self._concrete:
+            self._normal = False
```
Comment (scheibelp) [inline lib/spack/spack/spec.py:3211]: Side note: without this the old concretizer will "flatten" concrete already-installed specs but doesn't necessarily have the info to reassemble them, so we avoid disconnecting concrete sub-specs.

File: lib/spack/spack/spec.py [hunk 3]
```diff
@@ -3511,42 +3507,38 @@ def _dup(self, other, deps=True, cleardeps=True, caches=None):
...
             if isinstance(deps, (tuple, list)):
                 deptypes = deps
-            self._dup_deps(other, deptypes, caches)
+            self._dup_deps(other, deptypes)
 
         self._concrete = other._concrete
         self._hashes_final = other._hashes_final
 
-        if caches:
+        if self._concrete:
```
Comment (scheibelp) [inline lib/spack/spack/spec.py:3757]: The most important aspect of this logical change is that hashes on concrete specs are always preserved (in particular those read from the Cray manifest, which should not recompute their hash).

There might be times when retaining the `caches` parameter could avoid recomputing the hash, but this isn't expected to be a major issue (for one, it shouldn't add any cost when using the ASP concretizer).

**Review by tgamblin — APPROVED** [review]

**Pull request merged**